### ISSUE TITLE
Dynamic "tables"

### ISSUE THREAD

**Issue by CoskunSunali** [linq2db/linq2db#1522]
I have an application where administrators can design the database using a UI and this part of the application is handled without linq2db, meaning, it already works.

If some developer developed a typed entity, it works fine with linq2db but what if there is no typed entity?

Imagine I have a table that linq2db has no idea about. I would like to be able to query that table using the dynamic columns store feature.

```cs
class DynamicTable
{
    [PrimaryKey, Identity]
    public long Id { get; set; }

    [DynamicColumnsStore]
    public IDictionary<string, object> ExtendedProperties { get; set; }
}
```

That requires several things.

1. `Entity` method needs a `tableName` parameter or some unique value that we can use when building schema for multiple `DynamicTable` instances, because there can be more than one with each having different dynamic columns.

```cs
// Current:
Entity<DynamicTable>()

// Hoped:
Entity<DynamicTable>("Table1").Property... // (column1, column2)
Entity<DynamicTable>("Table2").Property... // (otherColumn1, otherColumn2)
```

2. `DataConnection.GetTable<T>` needs another overload there we can pass it the exact `DynamicTable` instance we want to construct.

```cs
// Current:
cnn.GetTable<DynamicTable>();

// Hoped:
cnn.GetTable<DynamicTable>("Table1")
cnn.GetTable<DynamicTable>("Table2")
```

3. Use fluent mapper to define dynamic columns in advance. linq2db doesn't know those but we do. There is one problem here, `Sql.Property` method is generic. It would be really helpful if there was a non-generic variation of it. So that we don't have to use reflection.

```cs
// Current:
Property(x => Sql.Property<string>(x, "DynColumn"))

// Hoped:
Property(x => Sql.Property(typeof(string), "DynColumn")) // Why do we need "x"?
```

Finally, the easiest of all of these is being able to implement a custom `IDynamicColumnMetadataProvider`, something similar to below. That would eliminate all of the requirements above (except GetTable<T>("name")). We wouldn't need to register dynamic columns in advance. You could easily ask us to provide you the list/schema of dynamic columns the first time that table is requested from the connection. I said "except GetTable<T>(name)" because the `name` parameter is used in the `IDynamicColumnMetadataProvider`.

```cs
class DynamicColumnMetadata
{
    public Type Type { get; set; }
    public Type Name { get; set; }
    public Attribute[] Attributes { get; set; }
}

interface IDynamicColumnMetadataProvider
{
    DynamicColumnMetadata[] GetDynamicColumns(string tableName);
}

class DynamicColumnMetadataProvider : IDynamicColumnMetadataProvider
{
    public DynamicColumnMetadata[] GetDynamicColumns(string tableName)
    {
        if (tableName == "Table1")
        {
            // return dynamic columns for Table1
        }
        else if (tableName == "Table2")
        {
            // return dynamic columns for Table2
        }
    }
}

// On initialize
var ms = new MappingSchema
{
    MetadataReader = new CustomMetadataReader(),
    DynamicColumnsMetadataReader = new CustomDynamicColumnsMetadataReader()
}

cnn.AddMappingSchema(ms);
```

What is your opinion on this use case? Is there a better approach that I am not aware of?

Thank you in advance!

**Comment by MaceWindu** [comment]
Some questions/suggestions

1. Can you use
```cs
class Entity1 : DynamicTable {}
class Entity2 : DynamicTable {}
```
or you cannot assign each table own type?

2. It is already possible using
```cs
db.GetTable<Table>().TableName("table_name")
```
take into account this issue #1274 and don't reuse GetTable result

3. Sql.Property is API for queries, that's why it is generic and needs x entity parameter. Probably we can add simplified fluent mapper API for dynamic properties to avoid confusion.

**Comment by CoskunSunali** [comment]
1. Unfortunately I don't have separate concrete implementation for each type, they are designed by "end users", not even developers. They can create new entity types using a UI.

2. This would help, I guess. Thanks!

3. That would be great. I assume I have to define each and every dynamic property using the fluent mapper otherwise they won't be filled into the dynamic columns store, right?

Finally, what do you think of `IDynamicColumnMetadataProvider` provided above?

**Comment by CoskunSunali** [comment]
By the way, I said "this would help" for 2nd clause but it wouldn't.

My initial `GetTable<T>("Table1")` example doesn't define a table name there. I mean, `Table1` is not the table name but the `entity mapping key` referring to `Entity<DynamicTable>("Table1")`.

Does that make sense?

**Comment by MaceWindu** [comment]
Regarding `IDynamicColumnMetadataProvider` - not sure why you want new interface added to linq2db. Custom IMetadataReader implementation should be enough or I miss something?

E.g. in my project I have custom table/column attributes and metadata reader, where (except other changes) I store table/column keys in attributes and metadata reader resolve them to actual database object names using lookup by provided key. But my case is easier, as I can pre-generate required table classes and don't need to use dynamic properties.

**Comment by CoskunSunali** [comment]
Just to clarify a bit further,

I would be doing the following if I wanted to define a dynamic table with a "key" and a table name.

```cs
Entity<DynamicTable>("Table1").HasTableName("Table1Name");
Entity<DynamicTable>("Table2").HasTableName("Table2Name");
```

So the initial `Table1` is not a table name but just a key so that I can use that exact mapping when I use `GetTable<DynamicTable>("Table1")`.

That is, in a way, saying, `I know there are multiple mappings for DynamicTable class but I want exactly that mapping`. Table name is another story :)

**Comment by sdanyliv** [comment]
@CoskunSunali, show me exmple how do you plan to use this in linq query.

**Comment by CoskunSunali** [comment]
> Regarding IDynamicColumnMetadataProvider - not sure why you want new interface added to linq2db. Custom IMetadataReader implementation should be enough or I miss something?

Custom `IMetadataReader` is only for providing mapping attributes. Can I use it for giving linq2db a list of dynamic columns?

I see the following:

```cs
        public MemberInfo[] GetDynamicColumns(Type type)
        {
            return Array.Empty<MemberInfo>();
        }
```

However, I thought it wanted me to return a list of "real" member info instances, like `PropertyInfo`. Do you have a class somewhere in linq2db that inherits from `MemberInfo` that I can use to return dynamic columns there?

E.g.: 

```cs
        public MemberInfo[] GetDynamicColumns(Type type)
        {
            var tableName = type.GetCustomAttribute<TableAttribute>();

            if (tableName == "Table1") {
                return new[] { new CustomMemberInfo("DynColumn1") }
            }
        }
```

So I am trying to figure if this method does exactly what I imagine of it and if there is a MemberInfo that I can use from linq2db.

**Comment by MaceWindu** [comment]
Check `FluentMetadataReader`'s implementation of GetDynamicColumns - maybe it will help more

**Comment by CoskunSunali** [comment]
@sdanyliv just like how I use dynamic columns normally. However, there is no concrete class implementation per table.

Imagine this, I have a database table `Tables` and another table `Columns`. The users can add tables and define their columns.

I have auto migrations that can change db schema based on these two tables. This is done automatically.

I finally have web parts that can query these data.

So the final query is always created at runtime.

The problem here is the concrete types missing at compile time and all tables using the same concrete type `DynamicTable` which has a primary key and a dynamic columns store. I want to fill it in based on the data I have in my `Tables` and `Columns` tables (querying these two is outside the scope of linq2db, I am aware.

**Comment by CoskunSunali** [comment]
So, I just tried returning an array of `DynamicColumnInfo` in my custom `IMetadataReader.GetDynamicColumns` and it really helped! Thanks guys!

There is of course the bad news, it is blocked by #1521 :)

What is more, in this case, I don't have to use fluent mapper to define my dynamic columns because I can do so using `GetDynamicColumns` method of my custom `IMetadataReader`. However, this has one down side that I cannot provide further information about the dynamic column (e.g. not nullable, length, etc).

That being said, the biggest problem here continues to be `not having separate concrete types for each table`...

**Comment by sdanyliv** [comment]
To define nullable, not nullable, etc, you have to return this information in other method
https://github.com/linq2db/linq2db/blob/b182372bc62367a134670212d0ae92df3c2f9b30/Source/LinqToDB/Metadata/MetadataReader.cs#L46

**Comment by CoskunSunali** [comment]
> @CoskunSunali, show me exmple how do you plan to use this in linq query.

@sdanyliv it is exactly the same as how one would query dynamic columns. The only difference is the `dynamic table` part.

```cs
var x = await cnn
    .GetTable<DynamicTable>("someKeyButNotTableName") // this is not the table name, but the mapping key
    .Table("ThisTimeTableName") // this is the table name
    .ToListAsync();
```

I would expect that to return the dynamic columns store filled with dynamic column values because the mapping already knows what dynamic fields I have, am I wrong?

**Comment by CoskunSunali** [comment]
> To define nullable, not nullable, etc, you have to return this information in other method
> [linq2db/Source/LinqToDB/Metadata/MetadataReader.cs](https://github.com/linq2db/linq2db/blob/b182372bc62367a134670212d0ae92df3c2f9b30/Source/LinqToDB/Metadata/MetadataReader.cs#L46)
> 
> Line 46 in [b182372](/linq2db/linq2db/commit/b182372bc62367a134670212d0ae92df3c2f9b30)
> 
>  public T[] GetAttributes<T>(Type type, MemberInfo memberInfo, bool inherit)

Actually that didn't help because the custom memberinfo implementation doesn't allow me to use `GetCustomAttribute` on it. System.Reflection throws some exception I cannot recall now. Do you want me to check?

**Comment by sdanyliv** [comment]
We should start here using TDD approach. You write test, and we fix everything or help you to write working solution.

**Comment by sdanyliv** [comment]
It is easy, just follow these steps
https://github.com/linq2db/linq2db/issues/1303#issuecomment-425113822

**Comment by CoskunSunali** [comment]
Sure, I will do that.

**Comment by sdanyliv** [comment]
Left your tests in Playground project and create PR, then we will handle this.

**Comment by CoskunSunali** [comment]
Thanks guys, great support as always, which is why I will be more than happy whenever I can contribute.

**Comment by CoskunSunali** [comment]
Do you have any idea why I would be getting the following exception?

```
[12.01.2019 10:15:39 Informational] ------ Discover test started ------
[12.01.2019 10:15:39 Informational] NUnit Adapter 3.10.0.21: Test discovery starting
[12.01.2019 10:15:39 Informational] Assembly contains no NUnit 3.0 tests: D:\work-external\github\CoskunSunali\linq2db\Source\LinqToDB\bin\Debug\net45\linq2db.dll
[12.01.2019 10:15:39 Informational] Assembly contains no NUnit 3.0 tests: D:\work-external\github\CoskunSunali\linq2db\Tests\Model\bin\Debug\net45\linq2db.Model.dll
[12.01.2019 10:15:40 Informational] NUnit Adapter 3.10.0.21: Test discovery complete
[12.01.2019 10:15:41 Informational] ========== Discover test finished: 59 found (0:00:02.1561238) ==========
[12.01.2019 10:15:56 Informational] ------ Run test started ------
[12.01.2019 10:15:56 Informational] NUnit Adapter 3.10.0.21: Test execution started
[12.01.2019 10:15:57 Informational] Running selected tests in D:\work-external\github\CoskunSunali\linq2db\Tests\Tests.Playground\bin\Debug\net462\linq2db.Tests.Playground.dll
[12.01.2019 10:15:57 Warning] Dependent Assembly Mono.Cecil, Version=0.9.6.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756 of D:\work-external\github\CoskunSunali\linq2db\Tests\Tests.Playground\bin\Debug\net462\linq2db.Tests.Playground.dll not found. Can be ignored if not a NUnit project.
[12.01.2019 10:15:57 Warning] Exception System.NullReferenceException, Exception thrown unloading tests from D:\work-external\github\CoskunSunali\linq2db\Tests\Tests.Playground\bin\Debug\net462\linq2db.Tests.Playground.dll
[12.01.2019 10:15:57 Warning] Object reference not set to an instance of an object.
[12.01.2019 10:15:57 Warning]    at NUnit.VisualStudio.TestAdapter.NUnit3TestExecutor.RunAssembly(String assemblyPath, TestFilter filter) in D:\repos\nunit\nunit3-vs-adapter\src\NUnitTestAdapter\NUnit3TestExecutor.cs:line 314
[12.01.2019 10:15:57 Informational] NUnit Adapter 3.10.0.21: Test execution complete
[12.01.2019 10:15:57 Informational] ========== Run test finished: 0 run (0:00:00.5166195) ==========
```

I just forked linq2db, copied `TestTemplate` into a new file `DynamicTableTests` and nothing else.

**Comment by sdanyliv** [comment]
Try run.
`dotnet restore inq2db.playground.sln`
But possibe `UserDataProviders.json` Improperly configured.

**Comment by MaceWindu** [comment]
probably you should update nunit adapter to 3.11 at least

**Comment by CoskunSunali** [comment]
I and @sdanyliv are working on a different way of implementing this so I am not writing any tests for the exact given case.

I would like to kindly ask @sdanyliv to decide whether he wants to close this issue or not.

Thank you both for your ideas and comments, it has been a pleasure to brainstorm with you.

**Comment by sdanyliv** [comment]
Closing for now, since wa are working fo real solution which works with dynamically created types,
Who are interested, there is branch `dynamic_querying`.
It is not ready for WIP PR, but you can examine what can be done with dynamic queries,

**Comment by vladimir-kovalyuk** [comment]
I had a look to the sources of dynamic_querying branch. Looks promising. When is it expected to be released? And where does the current discussion go?

**Comment by guillaume86** [comment]
I'm also interested in the dynamic querying feature, any news?

**Comment by BenKmann** [comment]
The Branch dynamic_querying seems to be outdatet sadly. Sounds like a nice Feature.

**Comment by junsan1** [comment]
Is it in 5.0.0?

**Comment by sdanyliv** [comment]
Not expected to release, since we have no updated requirements. I need to work close with developer which need that and probably API design.

**Comment by junsan1** [comment]
```
private class TestEntity
{
	public IDictionary<string, object> Extra { get; set; }
}

var entityBuilder = builder.Entity<TestEntity>();

// I can retrieve column names by my own logic
foreach (var colName in new[] { "Col1", "Col2", "Col3" })
{
	entityBuilder.Property(
		colName, // direct set column name

		// getter (simple logic, but can be extended (e.g. I can convert colName into some complex object in case Extra has non primitive key type) so Expression in that case - isn't a good option. Classic delegates do not has limitation like Expression
		(e, colName) => e.Extra[colName],

		// setter (some logic like getter and same possibilities
		(e, colName, rowValue) => e.Extra[colName] = rowValue
	);
}
```

**Comment by hmettmann** [comment]
I would also be very interested in this feature. I tried to make a sample use case below. If you have any questions, let me know: 

The idea is to have a user defined database-structure. The user can define which tables exists and which columns the table has at runtime. So my model has to be as flexible as well.
(I know I could use basic DataTables for that regard, but this flexible model is just a part of the application and the rest uses linq2db with normal models. It would be nice, if I could use it for the dynamic part as well)

My Idea is to use it like that:

```
// Generic Model Entity for all generic tables
public class DynamicEntity
{
  [PrimaryKey]
  public string Id { get; set; }
  
  [Column]
  [NotNull]
  public DateTime ChangeDateTime { get; set; }

  [DynamicColumnsStore]
  public Dictionary<string, object> FieldValues { get; set;} = new ();
}
```
Fluent Mapping Initialization:

```
var schema = new MappingSchema();
var builder = new FluentMappingBuilder(schema);

Dictionary<string, IEnumerable<string>> fieldsByTablename = LoadFieldConfig(); // The Field-Config is loaded here.
// Initialize the fluent mapping
foreach (var table in fieldsByTablename)
{
	string tableName = table.Key;
	var tableBuilder = builder.Entity<DynamicEntity>();
	tableBuilder.HasTableName(table.Key)
				 .HasPrimaryKey(x => x.Id);

	foreach (var field in table)
	{
		tableBuilder.Property(p => Sql.Property<string>(p, field.FieldName)).HasDataType(DataType.NVarChar);

		// ...an so on for ever necessary column-type
		//
	}
}
```

Usage

```
// Getting my entities for a given TableName but with the same generic Model class. 
// Thats the main Point here, to be able to use the same class for more than one table just by giving the tableName as a parameter
string tableName1 = "MyTestTable1";
var myEntities1 = dbContext.GetTable<DynamicEntity>(tableName1).ToList(); 
// As SQL it should Result in a statement like that: "SELECT Id, myDynamicColumn1, myDynamicColumn2, ... FROM MyTestTable1

string tableName2 = "MyTestTable2";
var myEntities2 = dbContext.GetTable<DynamicEntity>(tableName2).ToList();
```
